### ISSUE TITLE
salt --batch causes traceback "ValueError: list.remove(x): x not in list" (in 2015.8.1)

### ISSUE THREAD

**Issue by arthurlogilab** [saltstack/salt#29298]
```
root@salt:/tmp# salt --batch 10 \* test.ping
[snip]
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 454, in salt_main
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.7/dist-packages/salt/cli/batch.py", line 173, in run
    minion_tracker[queue]['minions'].remove(id)
ValueError: list.remove(x): x not in list
Traceback (most recent call last):
  File "/usr/bin/salt", line 10, in <module>
    salt_main()
  File "/usr/lib/python2.7/dist-packages/salt/scripts.py", line 454, in salt_main
    client.run()
  File "/usr/lib/python2.7/dist-packages/salt/cli/salt.py", line 98, in run
    for res in batch.run():
  File "/usr/lib/python2.7/dist-packages/salt/cli/batch.py", line 173, in run
    minion_tracker[queue]['minions'].remove(id)
ValueError: list.remove(x): x not in list
```

```
root@salt:/tmp# salt --versions-report
Salt Version:
           Salt: 2015.8.1
 
Dependency Versions:
         Jinja2: 2.7.3
       M2Crypto: Not Installed
           Mako: 1.0.0
         PyYAML: 3.11
          PyZMQ: 14.4.0
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: 3.5.0
       dateutil: 2.2
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
 
System Versions:
           dist: debian 8.2 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: debian 8.2 
 
```

This looks similar to #26254 (and the batch mode has maybe bugs in the history #24875 #24617 #25155)

It seems that like suggested by @rbjorklin  https://github.com/saltstack/salt/issues/26254#issuecomment-158940440 this problem arises when a minion 2015.5.* is targeted in the batch (but 2015.8 is not yet packaged for wheezy, but that's another issue).

**Comment by rbjorklin** [comment]
@arthurlogilab try running the master in debug mode and do:

    salt '*' test.ping

and see if you spot any messages like these:

    [ERROR   ] An extra return was detected from minion your-hostname-here, please verify the minion, this could be a replay attack

If you do see those kind of messages verify that you haven't accidentally duplicated one of your nodes and kept the minion_id and pki/minion folder.

**Comment by jfindlay** [comment]
@arthurlogilab, thanks for the report.  @rbjorklin, thanks for the extra information.  If that case is causing a hard stack trace, it seem that salt should be handling it better.

**Comment by JaseFace** [comment]
I'm seeing the same on 2015.8.3.  There is another bug somewhere on FBSD causing the minion the sometimes return duplicate responses that I haven't rooted out yet, but the end result is that there are always a handful of minions that give 'extra return'.  Basically I can't use batch mode because of this failure, so just putting my vote towards a fix here :+1:

**Comment by vladislav-jomedia** [comment]
I have the same issue!
You have my vote too.

**Comment by vladislav-jomedia** [comment]
I've found the bad minion, which sent multiple replays , by replacing  

minion_tracker[queue]['minions'].remove(id)
with
                               if id in minion_tracker[queue]['minions']:
                                        minion_tracker[queue]['minions'].remove(id)
                               else:
                                       print_cli('Invalid minion :{0}'.format(id))

The problem was that on the minion side I had 3 processes of salt-minion running

**Comment by JaseFace** [comment]
@vladislav-jomedia Great catch, this is indeed an issue of a minion being removed multiple times because of too many minions running.  This quick patch works well for me.  Do you mind doing a PR with it, unless someone believes there is a better way to fix it?

**Comment by jfindlay** [comment]
This seems good to me.  @vladislav-jomedia, will you send in a pull request to 2015.8?  Thanks.

**Comment by lorengordon** [comment]
So, the problem is duplicate items of the same name in the same list? Why not take the set() of the list?

**Comment by lorengordon** [comment]
It looks like you could do that [here](https://github.com/saltstack/salt/blob/v2015.8.1/salt/cli/batch.py#L139), with:

```python
                minion_tracker[new_iter]['minions'] = set(next_)
```

**Comment by JaseFace** [comment]
@lorengordon it's the inverse.  Some minions are returning multiple times, so any subsequent attempts to remove the single existing entry for that minion fail.

**Comment by tjuup** [comment]
See that this is still open. 
Below lines fixed it for me.

Salt Version:
           Salt: 2015.8.8.2
/usr/lib/python2.7/dist-packages/salt/cli/batch.py

170                         else:
171                             parts.update(part)
172                             for id in part.keys():
173  +                               try:
174                                     minion_tracker[queue]['minions'].remove(id)
175  +                               except Exception as err:
176  +                                   print_cli('Unable to remove id :{0}'.format(id))

**Comment by rallytime** [comment]
Hi everyone - this has been fixed by #32686, which is available in the 2015.8.9 and up for the 2015.8 branch releases and the 2016.3.x releases.
